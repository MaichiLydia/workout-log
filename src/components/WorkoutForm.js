import React, { Component } from "react";
import Form from "./Form/Form";
import FieldContainer from "./Form/FieldContainer";
import Button from "./Form/Button";
import uuid from "uuid";
import PropTypes from "prop-types";

const initialState = {
  duration: {
    value: "",
    error: ""
  },
  activityId: {
    value: "",
    error: ""
  },
  date: {
    value: "",
    error: ""
  }
};
class WorkoutForm extends Component {
  constructor(props) {
    super(props);
    this.state = initialState;
  }
  reset() {
    this.setState(initialState);
  }

  createWorkout = e => {
    e.preventDefault();
    const workout = {
      id: uuid.v4(),
      duration: this.state.duration.value,
      activityId: parseInt(this.state.activityId.value),
      date: this.state.date.value
    };

    const isDisabled = this.isDisabled();
    if (!isDisabled) {
      this.props.addWorkout(workout);
      this.reset();
      e.currentTarget.reset();
    }
  };

  handleChange = (inputName, inputValue, error = "") => {
    this.setState({
      [inputName]: {
        value: inputValue,
        error: error
      }
    });
  };

  isDisabled() {
    return (
      !this.state.duration.value ||
      this.state.duration.error ||
      !this.state.activityId.value ||
      this.state.activityId.error ||
      !this.state.date.value ||
      this.state.date.error
    );
  }

  render() {
    return (
      <div className="container">
        <Form
          title="Workout Log"
          text="Add your exercises here !"
          onSubmit={this.createWorkout}
        >
          <FieldContainer
            error={this.state.duration.error}
            htmlFor="duration"
            title="Duration"
          >
            <FieldContainer.Input
              id="duration"
              name="duration"
              type="time"
              min="00:01"
              required={true}
              onChange={this.handleChange}
            />
          </FieldContainer>
          <FieldContainer
            error={this.state.activityId.error}
            htmlFor="activityId"
            title="Exercise"
          >
            <FieldContainer.Select
              id="activityId"
              name="activityId"
              options={this.props.activities}
              required={true}
              onChange={this.handleChange}
            />
          </FieldContainer>
          <FieldContainer
            error={this.state.date.error}
            htmlFor="date"
            title="Date"
          >
            <FieldContainer.Input
              id="date"
              type="date"
              name="date"
              min="2000-01-01"
              required={true}
              onChange={this.handleChange}
            />
          </FieldContainer>

          <Button disabled={this.isDisabled()}>Add</Button>
        </Form>
      </div>
    );
  }
}

WorkoutForm.propTypes = {
  addWorkout: PropTypes.func.isRequired,
  activities: PropTypes.array.isRequired
};
export default WorkoutForm;
