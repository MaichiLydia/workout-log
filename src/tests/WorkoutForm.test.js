import React from "react";
import WorkoutForm from "../components/WorkoutForm"
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';


it('check if WorkoutItem receives the right values', () => {
  const props = {
        workouts:[{
          id: "1",
          duration:"10:10",
          "activityId": 2,
          "date": '2019-01-01'
        },{
          id: "2",
          duration:"11:10",
          "activityId": 4,
          "date": '2019-02-01'
        },{
          id: "3",
          duration:"10:13",
          "activityId": 2,
          "date": '2019-04-01'
        }],
        activities:[{   
              id: "1",
              name: "Bike"
          },
          {   
              id: "2",
              name: "Jog"
          },
          {   
              id: "3",
              name: "Run"
          },
          {   
              id: "4",
              name: "Swim"
          }
      ]

      }
  const workoutForm = mount(<WorkoutForm {...props}/>)
  expect(toJson(workoutForm)).toMatchSnapshot()
});
