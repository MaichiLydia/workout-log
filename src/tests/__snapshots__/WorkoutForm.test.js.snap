// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check if WorkoutItem receives the right values 1`] = `
<WorkoutForm
  activities={
    Array [
      Object {
        "id": "1",
        "name": "Bike",
      },
      Object {
        "id": "2",
        "name": "Jog",
      },
      Object {
        "id": "3",
        "name": "Run",
      },
      Object {
        "id": "4",
        "name": "Swim",
      },
    ]
  }
  workouts={
    Array [
      Object {
        "activityId": 2,
        "date": "2019-01-01",
        "duration": "10:10",
        "id": "1",
      },
      Object {
        "activityId": 4,
        "date": "2019-02-01",
        "duration": "11:10",
        "id": "2",
      },
      Object {
        "activityId": 2,
        "date": "2019-04-01",
        "duration": "10:13",
        "id": "3",
      },
    ]
  }
>
  <div
    className="container"
  >
    <Form
      onSubmit={[Function]}
      text="Add your exercises here !"
      title="Workout Log"
    >
      <div>
        <h1>
          Workout Log
        </h1>
        <p>
          Add your exercises here !
        </p>
        <form
          className="form"
          onSubmit={[Function]}
        >
          <FieldContainer
            error=""
            htmlFor="duration"
            title="Duration"
          >
            <div
              className="form__field-container"
            >
              <label
                className="form__field-label"
                htmlFor="duration"
              >
                Duration
              </label>
              <Input
                id="duration"
                min="00:01"
                name="duration"
                onChange={[Function]}
                required={true}
                type="time"
              >
                <input
                  className="form__input"
                  id="duration"
                  min="00:01"
                  name="duration"
                  onBlur={[Function]}
                  type="time"
                />
              </Input>
            </div>
          </FieldContainer>
          <FieldContainer
            error=""
            htmlFor="activityId"
            title="Exercise"
          >
            <div
              className="form__field-container"
            >
              <label
                className="form__field-label"
                htmlFor="activityId"
              >
                Exercise
              </label>
              <Select
                id="activityId"
                name="activityId"
                onChange={[Function]}
                options={
                  Array [
                    Object {
                      "id": "1",
                      "name": "Bike",
                    },
                    Object {
                      "id": "2",
                      "name": "Jog",
                    },
                    Object {
                      "id": "3",
                      "name": "Run",
                    },
                    Object {
                      "id": "4",
                      "name": "Swim",
                    },
                  ]
                }
                required={true}
              >
                <select
                  className="form__select"
                  id="activityId"
                  name="activityId"
                  onChange={[Function]}
                >
                  <option
                    className="form__option"
                    key="1"
                    value="1"
                  >
                    Bike
                  </option>
                  <option
                    className="form__option"
                    key="2"
                    value="2"
                  >
                    Jog
                  </option>
                  <option
                    className="form__option"
                    key="3"
                    value="3"
                  >
                    Run
                  </option>
                  <option
                    className="form__option"
                    key="4"
                    value="4"
                  >
                    Swim
                  </option>
                </select>
              </Select>
            </div>
          </FieldContainer>
          <FieldContainer
            error=""
            htmlFor="date"
            title="Date"
          >
            <div
              className="form__field-container"
            >
              <label
                className="form__field-label"
                htmlFor="date"
              >
                Date
              </label>
              <Input
                id="date"
                min="2000-01-01"
                name="date"
                onChange={[Function]}
                required={true}
                type="date"
              >
                <input
                  className="form__input"
                  id="date"
                  min="2000-01-01"
                  name="date"
                  onBlur={[Function]}
                  type="date"
                />
              </Input>
            </div>
          </FieldContainer>
          <Button
            disabled={true}
          >
            <button
              className="form__btn form__btn--disabled"
              disabled={true}
            >
              Add
            </button>
          </Button>
        </form>
      </div>
    </Form>
  </div>
</WorkoutForm>
`;
